src/bgw/job.c

```diff
@@ -193,77 +203,148 @@ ts_bgw_job_get_all(size_t alloc_size, MemoryContext mctx)
 }
 
 static void
-bgw_job_scan_one(int indexid, ScanKeyData scankey[], int nkeys, tuple_found_func tuple_found,
-				 tuple_filter_func tuple_filter, void *data, MemoryContext mctx, LOCKMODE lockmode,
-				 bool fail_if_not_found)
+init_scan_by_job_id(ScanIterator *iterator, int32 job_id)
 {
-	Catalog *catalog = ts_catalog_get();
-	ScannerCtx scanctx = {
-		.table = catalog_get_table_id(catalog, BGW_JOB),
-		.index = catalog_get_index(catalog, BGW_JOB, indexid),
-		.nkeys = nkeys,
-		.scankey = scankey,
-		.tuple_found = tuple_found,
-		.filter = tuple_filter,
-		.data = data,
-		.lockmode = lockmode,
-		.scandirection = ForwardScanDirection,
-		.result_mctx = mctx,
-	};
-
-	ts_scanner_scan_one(&scanctx, fail_if_not_found, "bgw job");
+	iterator->ctx.index = catalog_get_index(ts_catalog_get(), BGW_JOB, BGW_JOB_PKEY_IDX);
+	ts_scan_iterator_scan_key_init(iterator,
+								   Anum_bgw_job_pkey_idx_id,
+								   BTEqualStrategyNumber,
+								   F_INT4EQ,
+								   Int32GetDatum(job_id));
 }
 
-static inline void
-bgw_job_scan_job_id(int32 bgw_job_id, tuple_found_func tuple_found, tuple_filter_func tuple_filter,
-					void *data, MemoryContext mctx, LOCKMODE lockmode, bool fail_if_not_found)
+/* Lock a job tuple using an advisory lock. Advisory job locks are
+ *  used to lock the job row while a job is running to prevent a job from being
+ *  modified while in the middle of a run. This lock should be taken before
+ *  bgw_job table lock to avoid deadlocks.
+ *
+ * returns whether or not the lock was obtained (false return only possible if block==false)
+ */
+
+static bool
+lock_job(int32 job_id, LOCKMODE mode, JobLockLifetime lock_type, LOCKTAG *tag, bool block)
 {
-	ScanKeyData scankey[1];
+	/* Use a special pseudo-random field 4 value to avoid conflicting with user-advisory-locks */
+	TS_SET_LOCKTAG_ADVISORY(*tag, MyDatabaseId, job_id, 0);
 
-	ScanKeyInit(&scankey[0],
-				Anum_bgw_job_stat_pkey_idx_job_id,
-				BTEqualStrategyNumber,
-				F_INT4EQ,
-				Int32GetDatum(bgw_job_id));
-	bgw_job_scan_one(BGW_JOB_STAT_PKEY_IDX,
-					 scankey,
-					 1,
-					 tuple_found,
-					 tuple_filter,
-					 data,
-					 mctx,
-					 lockmode,
-					 fail_if_not_found);
+	return LockAcquire(tag, mode, lock_type == SESSION_LOCK, !block) != LOCKACQUIRE_NOT_AVAIL;
 }
 
```

Comment:
It would be good to ad a comment either here or at the top of the file about the advisory lock and when it is acquired and why.